[Staking AHM] Migrated fast-unstake, nomination-pools and delegated-staking from `staking` to `staking-async` for testing and benchmark

sigurpol opened the pull request:
## Why

As mentioned in #8077, we want to ensure that `fast-unstake`, `nomination-pools` and `delegated-staking` rely on `pallet-staking-async` and not on `pallet-staking` for tests and benchmarks.

Since the new staking-async pallet's interface is largely compatible with the old `staking` pallet and tests in these external pallets are relying on the abstract interface `sp-staking`, the number of changes is relatively small.

Some key differences:
- We no longer allow the nomination of an invalid or non-existent validator (#8436). This change breaks some tests and needs to be addressed. 
- Removal of slashing spans (#8316), which were referenced by the nomination pools' benchmarks.
- When pools are completely destroyed and all stake is withdrawn, the async staking pallet's kill() method is called, which emits the StakerRemoved event. Tests need to be updated to take that into account.
- Asset Hub runtime uses RelaychainDataProvider as BlockNumberProvider for nomination pools. Commission throttling needs proper relay chain block number setup in nomination pool benchmarks.

Driven-by improvement in the nomination pools:
- addition of automatic chilling logic to the ⁠unbond function. When a pool is being destroyed and the depositor is unbonding all their remaining stake, the pool automatically chills (stops nominating) before attempting to unbond. This ensures the unbond operation succeeds.
- addition of a new `staking-async-testing-utils` pallet
```
  staking-async/src/testing_utils.rs (source of truth)
      ↑
      ├── staking-async/src/benchmarking.rs (imports directly)
      └── staking-async-testing-utils/src/lib.rs (re-exports)
          ↑
          └── External pallets (use testing-utils crate)
```

sigurpol committed:
fast-unstake: migrate from staking to staking-async

With the migration to staking-async, some functional changes are needed
in the tests:
- Validator Nominations: Updated all nomination calls to
use valid validators (`VALIDATOR_PREFIX`) instead of self-nomination or
invalid targets
- Genesis Setup: Fixed the genesis configuration to
properly set up validators and nominators with valid targets
- Exposure Creation: Updated `create_exposed_nominator` to use the correct
staking-async API

The benchmarks were failing because they were trying to nominate to
accounts that weren't properly registered validators. The PR #8436
correctly added validation to prevent this (fixing a silent failure),
but this broke benchmarks that had invalid setups.

sigurpol committed:
nomination-pools: migrate from staking to staking-async

sigurpol committed:
delegated-staking: migrate from staking to staking-async

1. Updated `pallet_staking_async::Config` implementation:
 - Replaced `ConvertCurve<RewardCurve>` with `MockEraPayout`
 - Added `RcClientInterface = MockRcClient`

2.Added missing dependencies and mocks:
 - Added `pallet-staking-async-rc-client` as a dev-dependency in `Cargo.toml`
 - Created `MockEraPayout` implementing `EraPayout<Balance>` trait
 - Created `MockRcClient` implementing `RcClientInterface` trait

3. Fixed genesis configuration:
 - Updated stakers format from 4-tuple to 3-tuple (removed controller field)
 - Replaced `minimum_validator_count` with correct configuration
 - Fixed `invulnerables` field to use `Default::default()` instead of empty vec

4. Enhanced test validator setup:
 - Added additional validators (18, 19, 20, 21, 22) with proper balances to support pool integration
tests (validators need to exist now to be successfully nominated!)
 - Updated validator count to 6 to accommodate the additional validators
 - Ensured account 99 is not pre-configured as a validator to maintain test expectations

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience runtime_dev --bump patch'

sigurpol committed:
Fixed prdoc

sigurpol committed:
nomations-pools: removed unused set_active_era()

sigurpol committed:
nomination-pools: removed call to add_slashing_spans()

Since PR #8316 removed slashing spans from `pallet-staking-async`, the `num_slashing_spans` parameter in these extrinsics is now deprecated and has no effect. The benchmarks were artificially creating slashing spans to test the worst-case scenario, but since spans no longer exist, we can simply remove the calls to `add_slashing_spans`.

sigurpol committed:
nomination-pools: add async staking RC client and fix mock

The above changes add the async staking remote calls client dependency
and refactor the mock runtime implementation to use updated staking
interfaces.

sigurpol committed:
nomination-pools: fixed current_era comparison

sigurpol committed:
nomination-pools: added test validators for nomination pool testing

This commit:
✅ **Fixes**: `BadTarget` errors - now pools can successfully nominate
valid validators (see #8436)

❌ **Doesn't fix**: `InsufficientBond` errors - this is a separate issue
where the pool becomes an actual nominator and hits minimum bond
requirements when unbonding

The `InsufficientBond` error is probably a consequence of this fix
working correctly - now that the pool is actually nominating, it's
subject to nominator bond requirements.

sigurpol committed:
nomination-pools: handled StakerRemoved Event

When pools are completely destroyed and all stake is withdrawn, the
async staking pallet's `kill()` method is called, which emits the
`StakerRemoved` event. This event was not emitted by the old `staking`
pallet so tests need to be updated to keep the new event into account

sigurpol committed:
Updated prdoc

sigurpol committed:
nomination-pools: auto-chill destroying pools when depositor fully
unbonds

sigurpol committed:
nomination-pools: added pool depositor unbond lifecycle tests

The tests verify automatic chill behavior when depositor unbonds from a
destroying pool, checking scenarios both with and without other members.

sigurpol committed:
nomination-pools: use pallet_staking_reward_curve

sigurpol committed:
Added nomination pools benchmarking to Asset Hub Westend

kianenigma commented on substrate/frame/fast-unstake/src/mock.rs:
```diff
@@ -119,18 +124,42 @@ impl frame_election_provider_support::ElectionProvider for MockElection {
-impl pallet_staking::Config for Runtime {
+// Mock era payout
+pub struct MockEraPayout;
+impl pallet_staking_async::EraPayout<Balance> for MockEraPayout {
```
If you also want a more elegant version of this for tests, I suggest: 

```rust
impl<B, T: Get<(B, B)>> EraPayout<B> for T {
    fn era_payout(...) -> (B, B) { T::get() }
}
```

and then you can use 

```rust
parameter_types! {
    type EraPayout: (Balance, Balance) = (1000, 100) 
}
```

as a thought exercise. Only apply if you think it is better.

To me it looks nicer, but it is likely because I am too used to our rather convoluted conventions and use of `Get` :)

sigurpol replied:
Done

kianenigma commented on substrate/frame/fast-unstake/src/tests.rs:
```diff
@@ -806,12 +806,8 @@ mod on_idle {
-				0
-			)
-			.is_some());
+			let exposure = Staking::eras_stakers(BondingDuration::get() - 1, &VALIDATOR_PREFIX);
```
maybe nicer here is to just do:

```
assert!(T::Staking::is_exposed_in_era(VALIDATOR_PREFIX, BondingDuration::get() - 1))
```

It seems like the impl of `is_exposed_in_era` defines being exposed as:

* either you are a nominator behind an exposure
* or you are the associated validator of that exposure (even if your `exposure.own == 0`)

kianenigma commented on substrate/frame/nomination-pools/src/lib.rs:
```diff
@@ -2300,6 +2300,15 @@ pub mod pallet {
 
+			// If the pool is being destroyed and the depositor is unbonding all their stake,
+			// automatically chill the pool first to avoid InsufficientBond errors.
+			if bonded_pool.is_destroying_and_only_depositor(member.active_points()) &&
```
This is a useful addition, allowing you to destroy the pool without needing to manually call `chill`. But I didn't understand why it has to happen in the context of this PR?

sigurpol replied:
actually, this is something I wanted to discuss, glad you noticed already 😄  I've added it  because `pool_chill_e2e()` test was failing w/o  and I thought that the test looked legit and  such kind of change would kind of make sense

sigurpol committed:
Create and use proper validators in benchmarks instead of mock accounts

sigurpol committed:
Fix taplo

sigurpol committed:
nomination-pools: fix more benchmarking tests

sigurpol committed:
Add extra comments and improve prdoc

sigurpol committed:
Add staking-async-rc-client to nomination pools benchmarks

sigurpol committed:
Improve prdoc

sigurpol committed:
Fix taplo

sigurpol committed:
Replace mock RcClient with unit type for tests

sigurpol commented on substrate/frame/delegated-staking/src/mock.rs:
```diff
@@ -102,19 +103,42 @@ impl onchain::Config for OnChainSeqPhragmen {
+
+// Mock RC client interface
+pub struct MockRcClient;
+impl rc_client::RcClientInterface for MockRcClient {
```
Done

sigurpol committed:
Refactor mock era payout to use generic parameter type

sigurpol committed:
Remove unneeded WeightInfo trait bounds

sigurpol commented on substrate/frame/fast-unstake/src/mock.rs:
```diff
@@ -119,18 +124,42 @@ impl frame_election_provider_support::ElectionProvider for MockElection {
+	type CurrencyToVote = SaturatingCurrencyToVote;
+	type EraPayout = MockEraPayout;
+	type RcClientInterface = MockRcClient;
+	type WeightInfo = ();
```
Done

sigurpol committed:
Improve validator exposure check readability

sigurpol committed:
Fix prdoc

sigurpol committed:
Replace generic EraPayout with dedicated test types

sigurpol committed:
Initialize bonded eras and era stakes in test setup

sigurpol committed:
Attempt to fix CI

Ank4n reviewed (commented):
Just a few minor nits and suggestions—overall, great work!

sigurpol commented on substrate/frame/fast-unstake/src/mock.rs:
```diff
@@ -119,18 +125,40 @@ impl frame_election_provider_support::ElectionProvider for MockElection {
-#[derive_impl(pallet_staking::config_preludes::TestDefaultConfig)]
-impl pallet_staking::Config for Runtime {
+pub struct TestEraPayout;
+impl pallet_staking_async::EraPayout<Balance> for TestEraPayout {
```
Done

sigurpol committed:
Yet another attempt to fix CI

sigurpol committed:
Cleanup westend runtime and fix CI

sigurpol committed:
Added feature gate runtime-benchmarks in testing_utils

sigurpol committed:
Make CI umbrella job happy

sigurpol committed:
Trying to make CI happy...

sigurpol committed:
Fix .github/scripts/check-workspace.py

sigurpol committed:
Fix bunch of cargo checks

sigurpol committed:
Trying to fix dev-dependency issue

sigurpol committed:
Create staking-async-testing-utils pallet

The new pallet is meant to be used by pallet like delegated_staking,
fast_unstake and nomination_pools to properly setup staking-async
environment.

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Clippy

sigurpol committed:
Fix CI warnings

1. Documentation link error in staking-async-testing-utils: Fixed by changing the syntax from
[pallet_staking_async::EraPayout] to [EraPayout](pallet_staking_async::EraPayout)
2. Unused import and mutable variables in westend runtime tests: Removed the unused
frame_support::assert_ok import and changed the mutable variables to immutable since they weren't
being modified
3. Unused import warning in staking-async benchmarking: Removed the wildcard import and used
explicit imports only for the functions actually needed, eliminating the test-specific re-export
that was causing duplicate import warnings

sigurpol committed:
Cleanup

- Optimized delegated-staking Cargo.toml dependencies
- Simplified delegated-staking mock.rs annotations
- Cleaned up test-delegate-stake Cargo.toml features

sigurpol committed:
Revert "Cleanup"

sigurpol committed:
Remove code duplication around staking-async benchmarking

sigurpol committed:
Fix benchmarking for nomination-pools

sigurpol committed:
Fix prdoc

sigurpol committed:
Fix prdoc... for real

sigurpol committed:
Fix clippy vs westend vs on-chain-release-build flag

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Fix benchmark for staking-async

sigurpol committed:
Fix staking-async benchmark vs WAH Wasm

sigurpol committed:
fast-unstake: fix  the deprecated controller AccountId in mock.rs

sigurpol commented on substrate/frame/nomination-pools/benchmarking/src/inner.rs:
```diff
@@ -70,6 +70,25 @@ fn create_funded_user_with_balance<T: pallet_nomination_pools::Config>(
 }
 
+// Create a funded validator with the given balance and set it up as a validator
+fn create_validator<T: Config>(n: u32, balance: BalanceOf<T>) -> T::AccountId {
```
Done

sigurpol committed:
nomination-pools-benchmarking: fix benchmark around set_commission

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Fix Cargo.lock after merge

sigurpol committed:
staking-async-testing-utils: add fn to set current era

sigurpol commented on substrate/frame/nomination-pools/benchmarking/src/inner.rs:
```diff
@@ -442,12 +461,13 @@ mod benchmarks {
+		);
 		// Set the current era
-		pallet_staking::CurrentEra::<T>::put(EraIndex::max_value());
+		pallet_staking_async::CurrentEra::<T>::put(EraIndex::max_value());
```
done

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Fix regression after merging #9079

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Cleanup

sigurpol committed:
Fix tests in nomination-pool/test-delegate-stake

sigurpol committed:
Cleanup

sigurpol committed:
nomination-pools-benchmarking: addhHelper function to ensure an account has the specified total balance

sigurpol committed:
delegated-staking: cleanup

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol commented on substrate/frame/nomination-pools/test-delegate-stake/Cargo.toml:
```diff
@@ -32,9 +32,17 @@ pallet-bags-list = { workspace = true, default-features = true }
 log = { workspace = true, default-features = true }
 sp-tracing = { workspace = true, default-features = true }
+
+[features]
```
note: we inherit the dependency through pallet-staking-async (see its Cargo.toml:57)

sigurpol commented on substrate/frame/delegated-staking/Cargo.toml:
```diff
@@ -13,23 +13,25 @@ targets = ["x86_64-unknown-linux-gnu"]
 frame-support = { workspace = true }
 frame-system = { workspace = true }
 log = { workspace = true }
+pallet-balances = { workspace = true }
```
note: I've moved deps from dev-dependencies here  for the complete build graph to work correctly with `runtime-benchmarks` feature

kianenigma replied:
If they are only needed for benchmarks, what you want to do is to add `optional = true` here, and then enable them in `runtime-benchmarks` feature.

sigurpol commented on polkadot/runtime/westend/Cargo.toml:
```diff
@@ -71,8 +70,6 @@ pallet-meta-tx = { workspace = true }
 pallet-migrations = { workspace = true }
 pallet-mmr = { workspace = true }
 pallet-multisig = { workspace = true }
-pallet-nomination-pools = { workspace = true }
```
I've removed delegated-staking and nomination-pools from westend runtime. @kianenigma , @Ank4n  please check if it makes sense.

kianenigma commented on polkadot/runtime/westend/src/lib.rs:
```diff
@@ -773,7 +774,7 @@ impl pallet_staking::Config for Runtime {
 	type HistoryDepth = frame_support::traits::ConstU32<84>;
 	type MaxControllersInDeprecationBatch = MaxControllersInDeprecationBatch;
 	type BenchmarkingConfig = polkadot_runtime_common::StakingBenchmarkingConfig;
-	type EventListeners = (NominationPools, DelegatedStaking);
```
You are doing it for a few of the staking pallets already though. Nompools and delegated staking are correctly removed.

sigurpol replied:
I am planning to tackle it in in #9442 via a  followup PR

kianenigma commented on polkadot/runtime/westend/src/tests.rs:
```diff
@@ -160,13 +159,7 @@ mod remote_tests {
-				"NominationPools".into(),
-				"DelegatedStaking".into(),
-			],
+			pallets: vec!["Staking".into(), "System".into(), "Balances".into()],
```
whole test can be removed now.

kianenigma commented on substrate/bin/node/runtime/src/lib.rs:
```diff
@@ -3018,7 +3018,6 @@ mod benches {
 		[pallet_migrations, MultiBlockMigrations]
 		[pallet_mmr, Mmr]
 		[pallet_multisig, Multisig]
-		[pallet_nomination_pools, NominationPoolsBench::<Runtime>]
```
sdounds good.

kianenigma commented on substrate/frame/delegated-staking/src/mock.rs:
```diff
@@ -193,38 +210,34 @@ impl ExtBuilder {
 				(GENESIS_NOMINATOR_ONE, 1000),
 				(GENESIS_NOMINATOR_TWO, 2000),
+				// Additional validators for tests
+				(18, 10000),
```
probably good to define 

`const EXTRA_VALIDATORS: [] = ..`

kianenigma commented on substrate/frame/staking-async-testing-utils/src/lib.rs:
```diff
@@ -0,0 +1,209 @@
+};
+
+// STAKING_ID constant
+pub const STAKING_ID: LockIdentifier = *b"staking ";
```
Does this have to stay is a leftover? we already have a `mod testing_utils` in `staking-async`.

kianenigma reviewed (commented):
First pass looks good, a few things I would have to take a closer look with more time

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
fmt

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol committed:
Replace MockRcClient with ()

sigurpol commented on substrate/frame/fast-unstake/src/mock.rs:
```diff
@@ -119,18 +125,40 @@ impl frame_election_provider_support::ElectionProvider for MockElection {
+}
+
+// Mock RC client interface
+pub struct MockRcClient;
```
good point, done!

sigurpol committed:
Cleanup Westend tests

sigurpol committed:
Replaced inline validator definitions with constants

sigurpol committed:
Removed STAKING_ID from staking-async-testing-utils

sigurpol committed:
Unusued import in westend/src/tests.rs

sigurpol committed:
delegated-staking: better handling of deps in Cargo.toml

sigurpol committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

sigurpol commented:
I believe we could tackle re-enabling of check-runtime-migration (see #8547)  in the scope of this PR if we remove all deps on old staking pallet from Westend runtime. Re-enabling the check makes try-state on staking fails on Westend, see  as example https://github.com/paritytech/polkadot-sdk/actions/runs/17401852305/job/49396795142

```log
 [2025-09-02T11:29:09Z INFO  runtime::frame-support] 🩺 Running "Staking" try-state checks
[2025-09-02T11:29:09Z ERROR runtime] panicked at /__w/polkadot-sdk/polkadot-sdk/substrate/frame/staking/src/pallet/impls.rs:2481:35:
    called `Option::unwrap()` on a `None` value
thread 'main' panicked at cli/main.rs:326:10:
```

This is on the try_state code path, which is ran automatically as a part of try-runtime on-runtime-upgrade. There is a CLI flag to turn it off, but we should not do that in the CI actually if not as temporary hack

Ank4n committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

Ank4n committed:
Merge branch 'master' into sigurpol/migrate_pallets_to_staking_async

Ank4n committed:
fix test

sigurpol commented:
Replaced by  #12402 + #11940
